File: src/details.test.ts
import { describe, expect, test } from "bun:test";
import type { K8sObject, PodMetrics } from "./kubectl.ts";
import { detailModel, ingressRuleRows, jobOwnedByCronJob, podMountsPVC } from "./details.ts";

const noTop = new Map<string, PodMetrics>();

describe("ingressRuleRows", () => {
    test("flattens host/path/service/port", () => {
        const ing = {
            spec: {
                rules: [
                    {
                        host: "app.example.com",
                        http: { paths: [{ path: "/", backend: { service: { name: "web", port: { number: 80 } } } }] },
                    },
                ],
            },
        } as unknown as K8sObject;
        expect(ingressRuleRows(ing)).toEqual([["app.example.com", "/", "web", "80"]]);
    });

    test("uses * when host is omitted", () => {
        const ing = { spec: { rules: [{ http: { paths: [{ path: "/api" }] } }] } } as unknown as K8sObject;
        expect(ingressRuleRows(ing)[0][0]).toBe("*");
    });
});

describe("podMountsPVC", () => {
    const pod = {
        spec: { volumes: [{ persistentVolumeClaim: { claimName: "data" } }, { emptyDir: {} }] },
    } as unknown as K8sObject;
    test("true when a volume references the claim", () => {
        expect(podMountsPVC(pod, "data")).toBe(true);
    });
    test("false otherwise", () => {
        expect(podMountsPVC(pod, "other")).toBe(false);
    });
});

describe("jobOwnedByCronJob", () => {
    test("matches ownerReference kind+name", () => {
        const job = { metadata: { ownerReferences: [{ kind: "CronJob", name: "nightly" }] } } as unknown as K8sObject;
        expect(jobOwnedByCronJob(job, "nightly")).toBe(true);
        expect(jobOwnedByCronJob(job, "other")).toBe(false);
    });
});

describe("detailModel", () => {
    test("service builds an endpointPods section from its selector", () => {
        const svc = {
            metadata: { name: "api", namespace: "default" },
            spec: { type: "ClusterIP", clusterIP: "10.0.0.1", selector: { app: "api" }, ports: [{ port: 80, targetPort: 8080 }] },
        } as unknown as K8sObject;
        const model = detailModel("services", svc, false, noTop);
        expect(model.section?.type).toBe("endpointPods");
        expect(model.summary.find(([k]) => k === "Selector")?.[1]).toBe("app=api");
    });

    test("node builds a nodePods section", () => {
        const node = { metadata: { name: "node-1" }, status: { nodeInfo: { kubeletVersion: "v1.29.0" } } } as unknown as K8sObject;
        const model = detailModel("nodes", node, false, noTop);
        expect(model.section).toEqual({ type: "nodePods", title: "Pods on node", node: "node-1" });
    });

    test("unknown kinds still get a summary (no dead end)", () => {
        const crd = { metadata: { name: "x", namespace: "default" }, apiVersion: "example.com/v1" } as unknown as K8sObject;
        const model = detailModel("widgets", crd, false, noTop);
        expect(model.summary.length).toBeGreaterThan(0);
        expect(model.section).toBeUndefined();
    });

    test("pod summary injects live CPU/Mem from metrics", () => {
        const pod = { metadata: { name: "web" }, status: { phase: "Running" } } as unknown as K8sObject;
        const top = new Map<string, PodMetrics>([["web", { cpu: "5m", memory: "20Mi" }]]);
        const model = detailModel("pods", pod, false, top);
        expect(model.summary.find(([k]) => k === "CPU / Mem")?.[1]).toBe("5m / 20Mi");
    });
});

File: src/internal/model/details_test.go
package model

import (
	"reflect"
	"testing"
)

// Ported from src/details.test.ts, case for case.

var noTop = map[string]Usage{}

func summaryValue(m DetailModel, key string) string {
	for _, r := range m.Summary {
		if r.Key == key {
			return r.Value
		}
	}
	return ""
}

func TestIngressRuleRowsFlattens(t *testing.T) {
	ing := obj(t, `{"spec":{"rules":[{"host":"app.example.com","http":{"paths":[
		{"path":"/","backend":{"service":{"name":"web","port":{"number":80}}}}]}}]}}`)
	got := IngressRuleRows(ing)
	want := [][]string{{"app.example.com", "/", "web", "80"}}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("got %v, want %v", got, want)
	}
}

func TestIngressRuleRowsWildcardHost(t *testing.T) {
	ing := obj(t, `{"spec":{"rules":[{"http":{"paths":[{"path":"/api"}]}}]}}`)
	if got := IngressRuleRows(ing)[0][0]; got != "*" {
		t.Errorf("host = %q, want *", got)
	}
}

func TestPodMountsPVC(t *testing.T) {
	pod := obj(t, `{"spec":{"volumes":[
		{"persistentVolumeClaim":{"claimName":"data"}},{"emptyDir":{}}]}}`)
	if !PodMountsPVC(pod, "data") {
		t.Error("expected true for a mounted claim")
	}
	if PodMountsPVC(pod, "other") {
		t.Error("expected false for an unmounted claim")
	}
}

func TestJobOwnedByCronJob(t *testing.T) {
	job := obj(t, `{"metadata":{"ownerReferences":[
		{"apiVersion":"batch/v1","kind":"CronJob","name":"nightly","uid":"x"}]}}`)
	if !JobOwnedByCronJob(job, "nightly") {
		t.Error("expected true for the owning cronjob")
	}
	if JobOwnedByCronJob(job, "other") {
		t.Error("expected false for a different name")
	}
}

func TestDetailModelServiceBuildsEndpointPodsSection(t *testing.T) {
	svc := obj(t, `{"metadata":{"name":"api","namespace":"default"},
		"spec":{"type":"ClusterIP","clusterIP":"10.0.0.1","selector":{"app":"api"},
		"ports":[{"port":80,"targetPort":8080}]}}`)
	m := DetailModelFor("services", svc, false, noTop)
	if m.Section == nil || m.Section.Type != "endpointPods" {
		t.Fatalf("section = %+v", m.Section)
	}
	if got := summaryValue(m, "Selector"); got != "app=api" {
		t.Errorf("selector = %q", got)
	}
	if got := summaryValue(m, "Ports"); got != "80→8080/TCP" {
		t.Errorf("ports = %q", got)
	}
}

func TestDetailModelNodeBuildsNodePodsSection(t *testing.T) {
	node := obj(t, `{"metadata":{"name":"node-1"},
		"status":{"nodeInfo":{"kubeletVersion":"v1.29.0"}}}`)
	m := DetailModelFor("nodes", node, false, noTop)
	want := &Section{Type: "nodePods", Title: "Pods on node", Node: "node-1"}
	if !reflect.DeepEqual(m.Section, want) {
		t.Errorf("section = %+v, want %+v", m.Section, want)
	}
}

func TestDetailModelUnknownKindsStillGetASummary(t *testing.T) {
	crd := obj(t, `{"metadata":{"name":"x","namespace":"default"},"apiVersion":"example.com/v1"}`)
	m := DetailModelFor("widgets", crd, false, noTop)
	if len(m.Summary) == 0 {
		t.Error("unknown kind produced an empty summary — that is a dead end")
	}
	if m.Section != nil {
		t.Errorf("unknown kind should have no section, got %+v", m.Section)
	}
}

func TestDetailModelPodSummaryInjectsLiveMetrics(t *testing.T) {
	pod := obj(t, `{"metadata":{"name":"web"},"status":{"phase":"Running"}}`)
	top := map[string]Usage{"web": {CPU: "5m", Memory: "20Mi"}}
	m := DetailModelFor("pods", pod, false, top)
	if got := summaryValue(m, "CPU / Mem"); got != "5m / 20Mi" {
		t.Errorf("cpu/mem = %q", got)
	}
}

// Namespace-qualified metrics must win over the bare name: two namespaces may
// hold pods with the same name, and the bare map would show one pod's CPU on
// the other's row.
func TestDetailModelPodSummaryPrefersQualifiedMetrics(t *testing.T) {
	pod := obj(t, `{"metadata":{"name":"web","namespace":"prod"},"status":{"phase":"Running"}}`)
	top := map[string]Usage{
		"web":      {CPU: "1m", Memory: "1Mi"},
		"prod/web": {CPU: "9m", Memory: "9Mi"},
	}
	m := DetailModelFor("pods", pod, false, top)
	if got := summaryValue(m, "CPU / Mem"); got != "9m / 9Mi" {
		t.Errorf("cpu/mem = %q, want the namespace-qualified sample", got)
	}
}

func TestDetailModelWorkloadUsesItsSelector(t *testing.T) {
	dep := obj(t, `{"metadata":{"name":"api","namespace":"default"},
		"spec":{"replicas":3,"selector":{"matchLabels":{"app":"api","tier":"web"}}},
		"status":{"replicas":3,"readyReplicas":2}}`)
	m := DetailModelFor("deployments", dep, true, noTop)
	if m.Section == nil || m.Section.Type != "workloadPods" {
		t.Fatalf("section = %+v", m.Section)
	}
	if m.Section.Selector != "app=api,tier=web" {
		t.Errorf("selector = %q", m.Section.Selector)
	}
}
